File: cpp/stack/generate_parentheses_leetcode_22.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution{
  private:
    stack<char> st;
    vector<string> res;
    int n;

    string concat_stack(stack<char> stk){
      string s = "";
      stack<char> rvstk;
      while(!stk.empty()){
        rvstk.push(stk.top());
        stk.pop();
      }
      while(!rvstk.empty()){
        s += rvstk.top();
        rvstk.pop();
      }
      cout << s << ", ";
      return s;
    }       

    void backtrack(int openN, int closeN){
      if(openN == closeN && closeN == this->n){
        res.push_back(this->concat_stack(this->st));  
      } 

      if(openN < this->n){
        this->st.push('(');
        backtrack(openN+1, closeN);
        this->st.pop();
      }

      if(closeN < openN){
        this->st.push(')');
        backtrack(openN, closeN+1);
        this->st.pop();
      }
    }

  public:
    vector<string> generateParenthesis(int n) {
      this->n = n;
      this->backtrack(0, 0);
      return res;
    }
};

int main(){
  Solution sol;
  cout<<"[";
  int n;
  cin >> n;
  for(string s : sol.generateParenthesis(n)){
    cout << s << ",";
  }
  cout << "]";
}

File: rust/stack/generate_parentheses_leetcode_22.rs
struct Solution {}

impl Solution {
    fn new() -> Self {
        Solution {}
    }

    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut res: Vec<String> = Vec::new();
        fn backtrack(openN: i32, closeN: i32, n: i32, current: &mut String, res: &mut Vec<String>) {
            if (openN == closeN) && (closeN == n) {
                res.push(current.clone());
            }

            if openN < n {
                current.push('(');
                backtrack(openN + 1, closeN, n, current, res);
                current.pop();
            }

            if closeN < openN {
                current.push(')');
                backtrack(openN, closeN + 1, n, current, res);
                current.pop();
            }
        }
        let mut current: String = String::new();
        backtrack(0, 0, n, &mut current, &mut res);
        res
    }
}

fn main() {
    let ans = Solution::generate_parenthesis(3);
    println!("{:?}", ans);
}
